packages/components/src/mobile/keyboard-aware-flat-list/index.ios.js

```diff
@@ -154,7 +154,7 @@ export const KeyboardAwareFlatList = ( {
 			scrollEventThrottle={ 16 }
 			style={ style }
 		>
-			<FlatList { ...props } />
+			<FlatList { ...props } scrollEnabled={ false } />
```

Comment:
Does this disable list virtualization or were we never benefiting from virtualization due to the nesting beneath a `ScrollView`? That is not entirely clear to me. 

My understanding of the approach in #33170 is that it flattened the two scroll lists into a single, virtualized list.

Reply:
> Does this disable list virtualization or were we never benefiting from virtualization due to the nesting beneath a `ScrollView`? That is not entirely clear to me.

Good question. I understand that we never benefited from virtualization due to nesting on iOS. In any case, I did some tests to confirm it:

**Enable debug mode and use smaller values for virtualization:**
```patch
diff --git forkSrcPrefix/packages/components/src/mobile/keyboard-aware-flat-list/index.ios.js forkDstPrefix/packages/components/src/mobile/keyboard-aware-flat-list/index.ios.js
index 8b233612763814cdcf7b5d8e59c3d822e0741ac5..0ea1e8478ad2a7620699f8e5df5b70f5f9bbe400 100644
--- forkSrcPrefix/packages/components/src/mobile/keyboard-aware-flat-list/index.ios.js
+++ forkDstPrefix/packages/components/src/mobile/keyboard-aware-flat-list/index.ios.js
@@ -154,7 +154,13 @@ export const KeyboardAwareFlatList = ( {
 			scrollEventThrottle={ 16 }
 			style={ style }
 		>
-			<FlatList { ...props } scrollEnabled={ false } />
+			<FlatList
+				{ ...props }
+				debug
+				initialNumToRender={ 1 }
+				maxToRenderPerBatch={ 1 }
+				windowSize={ 1 }
+			/>
 		</AnimatedScrollView>
 	);
 };
```

https://github.com/WordPress/gutenberg/assets/14905380/ed806048-bd17-478c-962f-ade834534d31

The right bar shows what items are being rendered in the `FlatList` and the scroll position. As you can see, there are no changes which mean that all items are being rendered when the list is mounted. Hence, there's no virtualization.

However, if we render the `FlatList` outside the scroll view, you can see that the virtualization is working.

```patch
diff --git forkSrcPrefix/packages/components/src/mobile/keyboard-aware-flat-list/index.ios.js forkDstPrefix/packages/components/src/mobile/keyboard-aware-flat-list/index.ios.js
index 8b233612763814cdcf7b5d8e59c3d822e0741ac5..51ea3db87ef6135e8f1f2b517a058e1ae4d97930 100644
--- forkSrcPrefix/packages/components/src/mobile/keyboard-aware-flat-list/index.ios.js
+++ forkDstPrefix/packages/components/src/mobile/keyboard-aware-flat-list/index.ios.js
@@ -142,6 +142,16 @@ export const KeyboardAwareFlatList = ( {
 
 	const style = [ { flex: 1 }, scrollViewStyle ];
 
+	return (
+		<FlatList
+			{ ...props }
+			debug
+			initialNumToRender={ 1 }
+			maxToRenderPerBatch={ 1 }
+			windowSize={ 1 }
+		/>
+	);
+
 	return (
 		<AnimatedScrollView
 			automaticallyAdjustContentInsets={ false }
```


https://github.com/WordPress/gutenberg/assets/14905380/4e56efb8-9666-47b6-9143-82e19cfd20bd
 
On the other hand, Android virtualization is currently working.

```patch
diff --git forkSrcPrefix/packages/components/src/mobile/keyboard-aware-flat-list/index.android.js forkDstPrefix/packages/components/src/mobile/keyboard-aware-flat-list/index.android.js
index eccb80f3903e5d3443391a37b3783961d9c19717..dd4a9965a1ba80ce7870080e4df3dd89bbe9177c 100644
--- forkSrcPrefix/packages/components/src/mobile/keyboard-aware-flat-list/index.android.js
+++ forkDstPrefix/packages/components/src/mobile/keyboard-aware-flat-list/index.android.js
@@ -19,6 +19,10 @@ export const KeyboardAwareFlatList = ( { innerRef, onScroll, ...props } ) => {
 				ref={ innerRef }
 				onScroll={ scrollHandler }
 				{ ...props }
+				debug
+				initialNumToRender={ 1 }
+				maxToRenderPerBatch={ 1 }
+				windowSize={ 1 }
 			/>
 		</KeyboardAvoidingView>
 	);
```

https://github.com/WordPress/gutenberg/assets/14905380/78db26dc-febc-47c5-b7e1-79f1d02dd063

Reply:
> My understanding of the approach in #33170 is that it flattened the two scroll lists into a single, virtualized list.

That's interesting, I can't recall if lists were being actually flattened 🤔. Let me do a quick check and share the findings.

Reply:
> > My understanding of the approach in #33170 is that it flattened the two scroll lists into a single, virtualized list.
> 
> That's interesting, I can't recall if lists were being actually flattened 🤔. Let me do a quick check and share the findings.

With the changes introduced in https://github.com/WordPress/gutenberg/pull/48791, iOS no longer uses `KeyboardAwareScrollView` so https://github.com/WordPress/gutenberg/pull/33170 can't really be applied. In fact, I'm going to close the PR.

One approach to enable virtualization would be to unify the `ScrollView` props into the `FlatList`, this way we'd only have one list being rendered. I haven't explored the feasibility of this, but might be worth taking a deeper look in the spirit of improving the performance and probably more importantly, memory consumption to avoid memory pressure crashes on long posts. `FlatList` component inherits `ScrollView`, so I understand that in theory, this approach would be possible 🤔 . I'll open a separate GitHub issue as a follow-up.

https://github.com/WordPress/gutenberg/blob/5bf6aa089c67685da4aa738a45d30ed32786fb73/packages/components/src/mobile/keyboard-aware-flat-list/index.ios.js#L146-L158

Reply:
Thank you for starting the conversation and debugging about the current implementation on iOS and having this nested in a ScrollView! Hope we can tackle https://github.com/WordPress/gutenberg/issues/53265 soon.